sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -495,6 +495,23 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Eliminate no-op Projects
     case p @ Project(_, child) if sameOutput(child.output, p.output) => child
 
+    // The column of father project contains not deterministic function
+    // e.g Rand function. father project will be split to two project.
+    case h @ Project(fields, _: LeafNode) if !fields.forall(_.deterministic) =>
```

Comment:
I meant, in all the cases projectList contains non-deterministic expr, do we always need to split the project on all kinds of LeafNode?